ksonnet-gen/ksonnet/emit.go

```diff
@@ -85,60 +159,83 @@ func (root *root) emit(m *indentWriter) {
 	m.indent()
 
 	// Emit in sorted order so that we can diff the output.
+	groups := root.groups.toSortedSlice()
+	var groupNames []string
+	for _, g := range groups {
+		groupNames = append(groupNames, string(g.name))
+	}
+
 	for _, group := range root.groups.toSortedSlice() {
-		group.emit(m)
+		if err := group.emit(m); err != nil {
+			return err
+		}
 	}
 
 	m.writeLine("local hidden = {")
 	m.indent()
 
 	for _, hiddenGroup := range root.hiddenGroups.toSortedSlice() {
-		hiddenGroup.emit(m)
+		if err := hiddenGroup.emit(m); err != nil {
+			return err
+		}
 	}
 
 	m.dedent()
 	m.writeLine("},")
 
 	m.dedent()
 	m.writeLine("}")
+
+	return nil
 }
 
-func (root *root) addDefinition(
-	path kubespec.DefinitionName, def *kubespec.SchemaDefinition,
-) {
-	parsedName := path.Parse()
+func (root *root) addDefinition(path kubespec.DefinitionName, def *kubespec.SchemaDefinition) error {
+	parsedName, err := path.Parse()
+	if err != nil {
+		return fmt.Errorf("parse path: %v", err)
+	}
+
 	if parsedName.Version == nil {
-		return
+		return nil
+	}
+	apiObject, err := root.createAPIObject(parsedName, def)
+	if err != nil {
+		return fmt.Errorf("create api object: %v", err)
 	}
-	apiObject := root.createAPIObject(parsedName, def)
 
 	for propName, prop := range def.Properties {
 		pm := newPropertyMethod(propName, path, prop, apiObject)
 		apiObject.properties[propName] = pm
 
 		st := prop.Type
-		if isMixinRef(pm.ref) ||
+		if pm.ref.IsMixinRef() ||
 			(st != nil && *st == "array" && prop.Items.Ref != nil) {
 			typeAliasName := propName + "Type"
 			ta, ok := apiObject.properties[typeAliasName]
 			if ok && ta.kind != typeAlias {
-				log.Panicf(
-					"Can't create type alias '%s' because a property with that name already exists", typeAliasName)
+				return fmt.Errorf("Can't create type alias %q because a property with that name already exists", typeAliasName)
 			}
 
 			ta = newPropertyTypeAlias(typeAliasName, path, prop, apiObject)
 			apiObject.properties[typeAliasName] = ta
 		}
 	}
+
+	return nil
 }
 
-func (root *root) createAPIObject(
+func (r *root) createAPIObject(
 	parsedName *kubespec.ParsedDefinitionName, def *kubespec.SchemaDefinition,
-) *apiObject {
+) (*apiObject, error) {
+	dn, err := parsedName.Unparse(r.isLegacySchema)
+	if err != nil {
+		return nil, fmt.Errorf("unparse definition name: %v", err)
+	}
```

Comment:
Since `dn` is only used in the error, you might want to move this into the error block. (Though this doesn't apply, in my estimation, below in functions like `getAPIObjectHelper`.)